File: mihon/src/all/r2merge/src/eu/kanade/tachiyomi/extension/all/r2merge/Sites.kt
package eu.kanade.tachiyomi.extension.all.r2merge

internal enum class SiteId {
    Nhentai,
    HentaiRead,
    HentaiNexus,
    Hentai2Read,
    PandaChaika,
    EHentai,
    Hitomi,
    ;

    fun label(): String = when (this) {
        Nhentai -> "nhentai"
        HentaiRead -> "hentairead"
        HentaiNexus -> "hentainexus"
        Hentai2Read -> "hentai2read"
        PandaChaika -> "pandachaika"
        EHentai -> "ehentai"
        Hitomi -> "hitomi"
    }
}

internal const val CHAIKA_BASE = "https://panda.chaika.moe"
internal const val HITOMI_BASE = "https://hitomi.la"
internal const val HITOMI_CDN = "gold-usergeneratedcontent.net"
internal const val HITOMI_LTN = "https://ltn.$HITOMI_CDN"
internal const val EHENTAI_BASE = "https://e-hentai.org"
internal const val EXHENTAI_BASE = "https://exhentai.org"

internal fun hostOf(url: String): String {
    val match = Regex("""^https?://([^/?#]+)""", RegexOption.IGNORE_CASE).find(url.trim())
    return (match?.groupValues?.get(1) ?: "").lowercase().removePrefix("www.")
}

internal fun normalizeSite(value: String?): SiteId? {
    if (value.isNullOrBlank()) return null
    return when (value.trim().lowercase().replace(Regex("""\s+"""), "-")) {
        "nhentai", "nh", "n-hentai", "nhentai.net" -> SiteId.Nhentai
        "hentairead", "hr", "hentai-read", "hentairead.com" -> SiteId.HentaiRead
        "hentainexus", "hn", "nexus", "hentai-nexus", "hentainexus.com" -> SiteId.HentaiNexus
        "hentai2read", "h2r", "hentai-2-read", "hentai2read.com" -> SiteId.Hentai2Read
        "pandachaika", "panda-chaika", "chaika", "panda", "panda.chaika.moe",
        "chaika.moe",
        -> SiteId.PandaChaika
        "ehentai", "e-hentai", "eh", "exhentai", "ex", "e-hentai.org", "exhentai.org" -> SiteId.EHentai
        "hitomi", "hitomi.la" -> SiteId.Hitomi
        else -> null
    }
}

internal fun siteFromHost(host: String): SiteId? {
    val h = host.lowercase().removePrefix("www.")
    return when {
        h == "nhentai.net" || h.endsWith(".nhentai.net") -> SiteId.Nhentai
        h == "hentairead.com" || h.endsWith(".hentairead.com") ||
            h == "hencover.xyz" || h == "henread.xyz" -> SiteId.HentaiRead
        h == "hentainexus.com" || h.endsWith(".hentainexus.com") -> SiteId.HentaiNexus
        h == "hentai2read.com" || h.endsWith(".hentai2read.com") ||
            h.endsWith(".hentaicdn.com") -> SiteId.Hentai2Read
        h == "panda.chaika.moe" || h == "chaika.moe" || h.endsWith(".chaika.moe") -> SiteId.PandaChaika
        h == "e-hentai.org" || h == "exhentai.org" ||
            h.endsWith(".e-hentai.org") || h.endsWith(".exhentai.org") -> SiteId.EHentai
        h == "hitomi.la" || h.endsWith(".hitomi.la") ||
            h == HITOMI_CDN || h.endsWith(".$HITOMI_CDN") -> SiteId.Hitomi
        else -> null
    }
}

internal fun identifySite(url: String, explicit: String? = null): SiteId {
    normalizeSite(explicit)?.let { return it }
    siteFromHost(hostOf(url))?.let { return it }
    throw Exception(
        "Unknown chapter host \"${hostOf(url).ifBlank { url }}\". " +
            "Set source to nhentai, hentairead, hentainexus, hentai2read, " +
            "pandachaika, ehentai, or hitomi.",
    )
}

internal fun extractRemoteId(site: SiteId, url: String): String {
    val trimmed = url.trim()
    return when (site) {
        SiteId.Nhentai ->
            Regex("""/g/(\d+)""", RegexOption.IGNORE_CASE).find(trimmed)?.groupValues?.get(1)
                ?: Regex("""\b(\d{4,})\b""").find(trimmed)?.groupValues?.get(1)
                ?: throw Exception("Could not parse nhentai id from $trimmed")
        SiteId.HentaiRead ->
            Regex("""/hentai/([^/?#]+)""", RegexOption.IGNORE_CASE).find(trimmed)
                ?.groupValues?.get(1)
                ?.let { java.net.URLDecoder.decode(it, "UTF-8") }
                ?: java.net.URLDecoder.decode(lastPathSegment(trimmed), "UTF-8")
                    .ifBlank { throw Exception("Could not parse hentairead slug from $trimmed") }
        SiteId.HentaiNexus ->
            Regex("""/(?:view|read)/(\d+)""", RegexOption.IGNORE_CASE).find(trimmed)
                ?.groupValues?.get(1)
                ?: Regex("""\b(\d+)\b""").find(trimmed)?.groupValues?.get(1)
                ?: throw Exception("Could not parse hentainexus id from $trimmed")
        SiteId.Hentai2Read -> {
            val path = trimmed.replace(Regex("""[?#].*$"""), "").trimEnd('/')
                .replace(Regex("""^https?://[^/]+""", RegexOption.IGNORE_CASE), "")
                .trim('/')
            path.ifBlank { throw Exception("Could not parse hentai2read path from $trimmed") }
        }
        SiteId.PandaChaika ->
            Regex("""/archive/(\d+)""", RegexOption.IGNORE_CASE).find(trimmed)?.groupValues?.get(1)
                ?: Regex("""[?&]archive=(\d+)""", RegexOption.IGNORE_CASE).find(trimmed)
                    ?.groupValues?.get(1)
                ?: Regex("""\b(\d+)\b""").find(trimmed)?.groupValues?.get(1)
                ?: throw Exception("Could not parse pandachaika id from $trimmed")
        SiteId.EHentai -> {
            val gallery = Regex(
                """/g/(\d+)/([0-9a-f]+)""",
                RegexOption.IGNORE_CASE,
            ).find(trimmed)
            when {
                gallery != null -> "${gallery.groupValues[1]}/${gallery.groupValues[2]}"
                else -> Regex("""/g/(\d+)""", RegexOption.IGNORE_CASE).find(trimmed)
                    ?.groupValues?.get(1)
                    ?: throw Exception("Could not parse e-hentai gallery id/token from $trimmed")
            }
        }
        SiteId.Hitomi ->
            Regex("""/(?:galleries|reader)/(\d+)""", RegexOption.IGNORE_CASE)
                .find(trimmed)?.groupValues?.get(1)
                ?: Regex("""-(\d+)\.html?""", RegexOption.IGNORE_CASE).find(trimmed)
                    ?.groupValues?.get(1)
                ?: Regex("""/galleries/(\d+)\.js""", RegexOption.IGNORE_CASE).find(trimmed)
                    ?.groupValues?.get(1)
                ?: Regex("""\b(\d{3,})\b""").find(trimmed)?.groupValues?.get(1)
                ?: throw Exception("Could not parse hitomi id from $trimmed")
    }
}

internal fun canonicalUrl(site: SiteId, remoteId: String): String = when (site) {
    SiteId.Nhentai -> "https://nhentai.net/g/$remoteId/"
    SiteId.HentaiRead -> "https://hentairead.com/hentai/$remoteId/"
    SiteId.HentaiNexus -> "https://hentainexus.com/view/$remoteId"
    SiteId.Hentai2Read -> "https://hentai2read.com/${remoteId.trimStart('/')}/"
    SiteId.PandaChaika -> "$CHAIKA_BASE/archive/$remoteId"
    SiteId.EHentai -> ehentaiGalleryUrl(remoteId, EHENTAI_BASE)
    SiteId.Hitomi -> "$HITOMI_BASE/galleries/$remoteId.html"
}

internal fun refererForImage(url: String): String? = when (siteFromHost(hostOf(url))) {
    SiteId.Nhentai -> "https://nhentai.net/"
    SiteId.HentaiRead -> "https://hentairead.com/"
    SiteId.HentaiNexus -> "https://hentainexus.com/"
    SiteId.Hentai2Read -> "https://hentai2read.com/"
    SiteId.PandaChaika -> "$CHAIKA_BASE/"
    SiteId.EHentai -> {
        val host = hostOf(url)
        if (host.contains("exhentai")) "$EXHENTAI_BASE/" else "$EHENTAI_BASE/"
    }
    SiteId.Hitomi -> "$HITOMI_BASE/"
    null -> null
}

internal fun hentaiReadLanguageFromUrl(url: String): String? {
    val lang = Regex(
        """/hentai/[^/]+/([^/?#]+)/(?:p/\d+)?/?$""",
        RegexOption.IGNORE_CASE,
    ).find(url)?.groupValues?.get(1)?.lowercase()
    return if (lang == null || lang == "p" || lang.toIntOrNull() != null) null else lang
}

internal fun pageListUrl(site: SiteId, remoteId: String, originalUrl: String): String = when (site) {
    SiteId.Nhentai -> "https://nhentai.net/api/v2/galleries/$remoteId"
    SiteId.HentaiRead -> {
        val language = hentaiReadLanguageFromUrl(originalUrl) ?: "english"
        "https://hentairead.com/hentai/$remoteId/$language/p/1/"
    }
    SiteId.HentaiNexus -> "https://hentainexus.com/read/$remoteId"
    SiteId.Hentai2Read -> {
        val path = if (remoteId.contains("/")) remoteId else "$remoteId/1"
        "https://hentai2read.com/${path.trimStart('/')}/"
    }
    SiteId.PandaChaika -> "$CHAIKA_BASE/api?archive=$remoteId"
    SiteId.EHentai -> {
        val origin = if (hostOf(originalUrl).contains("exhentai")) EXHENTAI_BASE else EHENTAI_BASE
        ehentaiGalleryUrl(remoteId, origin)
    }
    SiteId.Hitomi -> "$HITOMI_LTN/galleries/$remoteId.js"
}

internal fun siteReferer(site: SiteId): String = when (site) {
    SiteId.Nhentai -> "https://nhentai.net/"
    SiteId.HentaiRead -> "https://hentairead.com/"
    SiteId.HentaiNexus -> "https://hentainexus.com/"
    SiteId.Hentai2Read -> "https://hentai2read.com/"
    SiteId.PandaChaika -> "$CHAIKA_BASE/"
    SiteId.EHentai -> "$EHENTAI_BASE/"
    SiteId.Hitomi -> "$HITOMI_BASE/"
}

internal fun ehentaiGalleryUrl(remoteId: String, origin: String): String {
    val parts = remoteId.trim('/').split('/')
    val path = if (parts.size >= 2) {
        "/g/${parts[0]}/${parts[1]}/"
    } else {
        "/g/${parts[0]}/"
    }
    return "${origin.trimEnd('/')}$path?nw=always"
}

internal fun chaikaDownloadUrl(download: String): String {
    val trimmed = download.trim()
    val absolute = when {
        trimmed.startsWith("http://", ignoreCase = true) ||
            trimmed.startsWith("https://", ignoreCase = true) -> trimmed
        else -> CHAIKA_BASE + if (trimmed.startsWith("/")) trimmed else "/$trimmed"
    }
    val noQuery = absolute.substringBefore('?').trimEnd('/')
    return if (noQuery.endsWith("/download")) "$noQuery/" else "$noQuery/download/"
}

internal fun isEHentaiHost(host: String): Boolean {
    val h = host.lowercase().removePrefix("www.")
    return h == "e-hentai.org" || h == "exhentai.org" ||
        h.endsWith(".e-hentai.org") || h.endsWith(".exhentai.org")
}

internal fun isChaikaLoopback(url: String): Boolean = url.startsWith("https://127.0.0.1/#") || url.startsWith("http://127.0.0.1/#")

private fun lastPathSegment(url: String): String {
    val cleaned = url.replace(Regex("""[?#].*$"""), "").trimEnd('/')
    return cleaned.substringAfterLast('/')
}

File: src/sources/r2-library/sites.ts
export type SiteId =
  | "nhentai"
  | "hentairead"
  | "hentainexus"
  | "hentai2read"
  | "pandachaika"
  | "ehentai"
  | "hitomi";

const ARCHIVE_SOURCES = new Set([
  "zip",
  "cbz",
  "archive",
  "r2",
  "folder",
  "dir",
  "file",
  "r2-library",
]);

export const isAbsoluteHttpUrl = (url: string): boolean =>
  url.startsWith("http://") || url.startsWith("https://");

export const isRemoteArchiveUrl = (url: string): boolean => {
  const path = url.trim().split("?")[0]?.split("#")[0]?.toLowerCase() ?? "";
  return path.endsWith(".zip") || path.endsWith(".cbz");
};

export const hostOf = (url: string): string => {
  const match = /^https?:\/\/([^/?#]+)/i.exec(url.trim());
  return (match?.[1] ?? "").toLowerCase().replace(/^www\./, "");
};

export const normalizeSite = (value?: string | null): SiteId | null => {
  if (!value?.trim()) return null;
  const key = value.trim().toLowerCase().replace(/\s+/g, "-");
  switch (key) {
    case "nhentai":
    case "nh":
    case "n-hentai":
    case "nhentai.net":
      return "nhentai";
    case "hentairead":
    case "hr":
    case "hentai-read":
    case "hentairead.com":
      return "hentairead";
    case "hentainexus":
    case "hn":
    case "nexus":
    case "hentai-nexus":
    case "hentainexus.com":
      return "hentainexus";
    case "hentai2read":
    case "h2r":
    case "hentai-2-read":
    case "hentai2read.com":
      return "hentai2read";
    case "pandachaika":
    case "panda-chaika":
    case "chaika":
    case "panda":
    case "panda.chaika.moe":
    case "chaika.moe":
      return "pandachaika";
    case "ehentai":
    case "e-hentai":
    case "eh":
    case "exhentai":
    case "ex":
    case "e-hentai.org":
    case "exhentai.org":
      return "ehentai";
    case "hitomi":
    case "hitomi.la":
      return "hitomi";
    default:
      return null;
  }
};

export const siteFromHost = (host: string): SiteId | null => {
  const h = host.toLowerCase().replace(/^www\./, "");
  if (h === "nhentai.net" || h.endsWith(".nhentai.net")) return "nhentai";
  if (
    h === "hentairead.com" ||
    h.endsWith(".hentairead.com") ||
    h === "hencover.xyz" ||
    h === "henread.xyz"
  ) {
    return "hentairead";
  }
  if (h === "hentainexus.com" || h.endsWith(".hentainexus.com")) {
    return "hentainexus";
  }
  if (
    h === "hentai2read.com" ||
    h.endsWith(".hentai2read.com") ||
    h.endsWith(".hentaicdn.com")
  ) {
    return "hentai2read";
  }
  if (h === "panda.chaika.moe" || h === "chaika.moe" || h.endsWith(".chaika.moe")) {
    return "pandachaika";
  }
  if (
    h === "e-hentai.org" ||
    h === "exhentai.org" ||
    h.endsWith(".e-hentai.org") ||
    h.endsWith(".exhentai.org")
  ) {
    return "ehentai";
  }
  if (
    h === "hitomi.la" ||
    h.endsWith(".hitomi.la") ||
    h === "gold-usergeneratedcontent.net" ||
    h.endsWith(".gold-usergeneratedcontent.net")
  ) {
    return "hitomi";
  }
  return null;
};

export const isArchiveSource = (value?: string | null): boolean => {
  const v = value?.trim().toLowerCase().replace(/\s+/g, "-") ?? "";
  return ARCHIVE_SOURCES.has(v);
};

export const tryIdentifySite = (
  url: string,
  explicit?: string | null,
): SiteId | null => {
  if (isArchiveSource(explicit)) return null;
  const named = normalizeSite(explicit);
  if (named) return named;
  if (!url || isRemoteArchiveUrl(url)) return null;
  return siteFromHost(hostOf(url));
};

export const extractRemoteId = (site: SiteId, url: string): string => {
  const trimmed = url.trim();
  switch (site) {
    case "nhentai":
      return (
        /\/g\/(\d+)/i.exec(trimmed)?.[1] ??
        /\b(\d{4,})\b/.exec(trimmed)?.[1] ??
        ""
      );
    case "hentairead": {
      const slug = /\/hentai\/([^/?#]+)/i.exec(trimmed)?.[1];
      if (slug) return decodeURIComponent(slug);
      const last = trimmed.replace(/[?#].*$/, "").replace(/\/+$/, "").split("/").pop();
      return last ? decodeURIComponent(last) : "";
    }
    case "hentainexus":
      return (
        /\/(?:view|read)\/(\d+)/i.exec(trimmed)?.[1] ??
        /\b(\d+)\b/.exec(trimmed)?.[1] ??
        ""
      );
    case "hentai2read":
      return trimmed
        .replace(/[?#].*$/, "")
        .replace(/\/+$/, "")
        .replace(/^https?:\/\/[^/]+/i, "")
        .replace(/^\/+/, "");
    case "pandachaika":
      return /\/archive\/(\d+)/i.exec(trimmed)?.[1] ?? /\b(\d+)\b/.exec(trimmed)?.[1] ?? "";
    case "ehentai": {
      const match = /\/g\/(\d+)\/([0-9a-f]+)/i.exec(trimmed);
      return match ? `${match[1]}/${match[2]}` : "";
    }
    case "hitomi":
      return (
        /galleries\/(\d+)/i.exec(trimmed)?.[1] ??
        /\/(\d+)\.html/i.exec(trimmed)?.[1] ??
        /\b(\d{4,})\b/.exec(trimmed)?.[1] ??
        ""
      );
    default:
      return "";
  }
};

export const canonicalUrl = (site: SiteId, id: string): string => {
  switch (site) {
    case "nhentai":
      return `https://nhentai.net/g/${id}/`;
    case "hentairead":
      return `https://hentairead.com/hentai/${encodeURI(id)}/`;
    case "hentainexus":
      return `https://hentainexus.com/view/${id}`;
    case "hentai2read":
      return `https://hentai2read.com/${id.replace(/^\/+/, "")}/`;
    case "pandachaika":
      return `https://panda.chaika.moe/archive/${id}`;
    case "ehentai": {
      const [gid, token] = id.split("/");
      return token
        ? `https://e-hentai.org/g/${gid}/${token}/`
        : `https://e-hentai.org/g/${gid}/`;
    }
    case "hitomi":
      return `https://hitomi.la/galleries/${id}.html`;
    default:
      return id;
  }
};

export const cloudflareResolveUrl = (url: string): string | undefined => {
  switch (tryIdentifySite(url)) {
    case "hentairead":
      return "https://hentairead.com/hentai/?sortby=new";
    case "hentainexus":
      return "https://hentainexus.com/";
    case "hentai2read":
      return "https://hentai2read.com/";
    case "nhentai":
      return "https://nhentai.net/";
    default:
      return undefined;
  }
};

export const refererForImage = (url: string): string | undefined => {
  const site = siteFromHost(hostOf(url));
  switch (site) {
    case "nhentai":
      return "https://nhentai.net/";
    case "hentairead":
      return "https://hentairead.com/";
    case "hentainexus":
      return "https://hentainexus.com/";
    case "hentai2read":
      return "https://hentai2read.com/";
    case "pandachaika":
      return "https://panda.chaika.moe/";
    case "ehentai":
      return hostOf(url).includes("exhentai")
        ? "https://exhentai.org/"
        : "https://e-hentai.org/";
    case "hitomi":
      return "https://hitomi.la/";
    default:
      return undefined;
  }
};
